src/errors.rs

```diff
@@ -23,7 +23,7 @@ pub enum InternalError {
     CallingApplicationMistake(#[from] CallerError),
     #[error("Serialization Error")]
     Serialization,
-    #[error("Protocol error")]
+    #[error("Some player sent a message which does not match the protocol specification")]
     ProtocolError,
     #[error("Could not successfully generate proof")]
     CouldNotGenerateProof,
```

Comment:
There's some work to do here in determining which (if any) of these are user error.
Also, the code that uses this error type makes ample use of the `warn! `macro to add context. If bugs here aren't user error, then these should be `error!` instead, however, making this change will result in relentless terminal spam upon running `cargo make ci`

Reply:
I almost want to say that ideally, at least for logging, there could be two types of InternalInvariantFailed: one which can be hit by test cases (such as by passing weird inputs to functions that shouldn't normally happen, like the current code that throws CouldNotGenerateProof errors), and one that should never be hit no matter what and will yell at you in angry red text if it is (like the current `error!` macro does).

So like four logging levels total:
info/debug: benign information about program flow
warn: not great but not necessarily catastrophic (detected an incorrect input from another player)
error: catastrophic, execution will not finish, something is wrong and it's probably our fault
ERROR: angry red text that should never appear, even in tests

But I'm not sure where to put "the were enough incorrect inputs from other parties that termination is no longer possible" in this taxonomy. Maybe each detected fault is a warn! and if the progam has to stop, it's an error! (generated at some higher level in the code)?

Thoughts?

Reply:
Actually, thinking about this some more, there shouldn't be that many ERRORs in practice and most InternalInvariantFaileds should be reachable by giving functions weird inputs. So maybe what we need is a way to mark a specific test as "this should fail" such that it doesn't print error! logs (but otherwise would, so that you see them in tests that shouldn't fail).

Reply:
Re: terminal spam. I thought that the tests only print stderr logs when they fail? For example, we don't see the seeds that every test prints out unless the test fails. Why does this logging behave differently? @gatoWololo 

I agree with your last comment that we should assume any IIF is reachable by some weird combination of circumstances (even if that involves messing with internal state or storage or something), and I'd rather not try to distinguish between "things we haven't figured out how to test" vs "things that should actually never happen". The latter case feels more like something we should design better abstractions around so it's not even an option in the code.

Reply:
@marsella To give some context:
Cargo test attempts to intercept stdout and stderr. However, it is hard to do this across all OSs. So the implementation is rather poor and only catches straightforward things like calls to `println!` and `eprintln!`. Which is what our seed does:
```rust
eprintln!("seed: {seed:?}");
```
So this output is suppressed unless our test fails.

This doesn't work for our logging however, so the output leaks through and always shows up. Tom and I realized this a little bit ago, but didn't seem like a priority, (it took me a while to figure out this is what was happening!). So here we are now :face_exhaling: 

A couple of thoughts/opinions from me:
- For what its worth, I think the focus should be to have good error logging and reporting when actually using the library. And worry less about how logging behaves in Ci or tests.
- Currently, we enable logging on all tests by having every test set up logging at the very begging. An EZ:tm: solution would just be to turn off logging on our unit tests. The disadvantage here is that we would lose logging for tests (which only matters if a test fails and we want more info on "why?"). But if a test fails, and we really want the logging for further debugging, we could always edit that specific test initialize logging and re-run it. Which I think would be good enough? (Theoretically, we could end up with an irreproducible test, so we are not able to run it again with logging, but I don't think this edge case is worth worrying about).
- In the future we could consider writing logs from unit tests to a file, that way it doesn't spam the terminal and we could always check the file if needed, (for CI we would need to make sure the file is persisted after the run). This seemed low-priority (and unfun) to implement, so I haven't proposed it :sweat_smile: but "it shouldn't be too difficult..."

Reply:
Ah, I didn't realize the catching of of stderr is flaky.

I think the EZ approach sounds perfect. It matches our current setup, where if a test fails you can go do some manual poking (fix the seed) to get more information. Right now the only source of randomness in the library is the RNG (we don't have any external parties or network flakiness or whatever to cause other types of unpredictability), so irreproducible tests are, I think, not a concern at this stage. When we do add such tests, we could consider a different logging setup for them (since I think they'd be part of e2e tests that would live in a separate directory from all our unit tests).

With this approach, we can switch all the error logs to be `error!` and modify `init_testing` to not log errors by default (and document / implement separately how to log them if you need to). Tom, it might be appropriate to put that into a separate issue along with making decisions about this whole error type (that is, we'd leave this as-is for now and address it in a separate issue).

Reply:
Great, this approach sounds good to me!